src/PositionManager.sol

```diff
@@ -168,67 +168,56 @@ contract PositionManager is IPositionManager, ERC721Permit, PoolInitializer, Mul
         return abi.encode(currencyDelta);
     }
 
-    function burn(uint256 tokenId, address sender) internal {
-        _requireApprovedOrOwner(tokenId, sender);
+    /// note: this is overloaded with ERC721Permit._burn
+    function _burn(bytes memory params, address sender) internal returns (bytes memory) {
+        (uint256 tokenId, PoolPosition memory poolPos, bytes memory hookData) =
+            abi.decode(params, (uint256, PoolPosition, bytes));
+
+        _beforeModify(poolPos, tokenId, sender);
+        uint256 liquidity = _beforeBurn(poolPos, tokenId);
 
-        // Checks that the full position's liquidity has been removed and all tokens have been collected from tokensOwed.
-        _validateBurn(tokenId);
+        // Can only call modify if there is non zero liquidity.
+        BalanceDelta delta;
+        if (liquidity > 0) (delta,) = _modifyLiquidity(poolPos, -(liquidity.toInt256()), bytes32(tokenId), hookData);
 
-        delete tokenRange[tokenId];
+        delete poolPositions[tokenId];
         // Burn the token.
         _burn(tokenId);
+        return abi.encode(delta);
     }
 
-    function _modifyLiquidity(LiquidityRange memory range, int256 liquidityChange, bytes32 salt, bytes memory hookData)
+    function _beforeModify(PoolPosition memory poolPos, uint256 tokenId, address sender) private view {
+        if (!_isApprovedOrOwner(sender, tokenId)) revert NotApproved(sender);
+        if (poolPositions[tokenId] != poolPos.toId()) revert IncorrectPoolPositionForTokenId(tokenId);
+    }
+
+    function _modifyLiquidity(PoolPosition memory poolPos, int256 liquidityChange, bytes32 salt, bytes memory hookData)
         internal
         returns (BalanceDelta liquidityDelta, BalanceDelta totalFeesAccrued)
```

Comment:
every call to this function only uses the first return parameter... so this internal function should probably just return 1 of them